File: src/main/kotlin/com/wynnlab/spells/mage/IceSnake.kt
package com.wynnlab.spells.mage

import com.wynnlab.api.normalizeOnXZ
import com.wynnlab.spells.Spell
import com.wynnlab.spells.SpellData
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.Particle
import org.bukkit.Sound
import org.bukkit.entity.Mob
import org.bukkit.entity.Player
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType

object IceSnake : Spell(21, SpellData.METEOR) {
    private val hit: MutableList<Mob> = mutableListOf()
    private lateinit var iceLoc: Location

    override fun init() {
        iceLoc = player.location.clone().add(0.0, 1.0, 0.0)
    }

    override fun tick() {
        iceLoc.add(player.eyeLocation.direction.clone().normalizeOnXZ())
        val iceBlock = player.world.spawnFallingBlock(
            iceLoc,
            if (clone) Material.OBSIDIAN.createBlockData() else Material.PACKED_ICE.createBlockData()
        )
        iceBlock.dropItem = false
        iceBlock.setHurtEntities(false)
        player.spawnParticle(Particle.FIREWORKS_SPARK, iceLoc, 1, 1.0, 1.0, 1.0, 0.1)
        player.spawnParticle(Particle.BLOCK_CRACK, iceLoc, 9, 1.0, 1.0, 1.0, 1.0,
            if (clone) Material.OBSIDIAN.createBlockData() else Material.ICE.createBlockData()
        )
        player.playSound(iceLoc,
            if (clone) Sound.BLOCK_STONE_PLACE else Sound.BLOCK_GLASS_BREAK, if (clone) 2f else 1f, 0.75f
        )
        player.playSound(iceLoc,
            if (clone) Sound.ENTITY_WITHER_BREAK_BLOCK else Sound.BLOCK_FIRE_EXTINGUISH,
            if (clone) 0.25f else 0.5f, if (clone) 1.5f else 1f
        )
        for (e in iceLoc.getNearbyEntities(1.0, 1.0, 1.0)) {
            if (e is Player) {
                continue
            }
            if (e is Mob) {
                e.addPotionEffect(PotionEffect(PotionEffectType.SLOW, 300, 4))
                if (hit.contains(e)) {
                    continue
                }
                hit.add(e)
                e.damage(1.0, player)
            }
        }
    }
}

File: src/main/resources/classes/mage/ice_snake.py
from org.bukkit import Material, Particle, Sound
from org.bukkit.potion import PotionEffect, PotionEffectType

from com.wynnlab.spells import PySpell
from com.wynnlab.util import BukkitUtils

class Spell(PySpell):
    def __init__(self):
        self.hit = set()
        self.ice_loc = None

    def init(self):
        self.ice_loc = self.player.getLocation().clone().add(0, 1, 0)

    def tick(self):
        self.ice_loc.add(BukkitUtils.normalizeOnXZ(self.player.getEyeLocation().getDirection()))

        ice_block = self.player.getWorld().spawnFallingBlock(self.ice_loc, Material.OBSIDIAN.createBlockData() if self.clone else Material.PACKED_ICE.createBlockData())
        ice_block.setDropItem(False)
        ice_block.setHurtEntities(False)

        self.particle(self.ice_loc, Particle.FIREWORKS_SPARK, 1, 1, 1, 1, .1)
        self.particle(self.ice_loc, Particle.BLOCK_CRACK, 9, 1, 1, 1, 1, Material.OBSIDIAN.createBlockData() if self.clone else Material.ICE.createBlockData())
        self.sound(self.ice_loc, Sound.BLOCK_STONE_PLACE if self.clone else Sound.BLOCK_GLASS_BREAK, 2 if self.clone else 1, .75)
        self.sound(self.ice_loc, Sound.ENTITY_WITHER_BREAK_BLOCK if self.clone else Sound.BLOCK_FIRE_EXTINGUISH, .25 if self.clone else .5, 1.5 if self.clone else 1)

        for e in self.nearbyMobs(self.ice_loc, 1, 1, 1):
            if e in self.hit:
                continue
            self.hit.add(e)

            e.addPotionEffect(PotionEffect(PotionEffectType.SLOW, 300, 4))
            self.damage(e, False, .7, .5, 0, 0, .5, 0, 0)
            PySpell.knockback(e, VectorUP, .5)
